### ISSUE TITLE
NGINX reverse proxy configuration not working as expected ( remote-server setup / local VM installation )

### ISSUE THREAD

**Issue by Dave-At-GitHub** [matrix-org/synapse#2861]
### Description

I've installed matrix-synapse on a server ( Ubuntu Server 16.04 LTS with a fixed IP ) which is accessed via a reverse proxy ( NGINX ).
The installation + reverse proxy configuration is done as described at https://github.com/matrix-org/synapse.
Now there is following issue, if I visit the server at port 8008 the matrix web interface is shown ( URL changes to 'http://domain-name:8008/_matrix/client/#/login' but if I visit 'http://reverse-proxy-domain-name' ( server name configured in NGINX ) only the welcome to NGINX page is shown.

So for testing purposes I've installed matrix-synapse + NGINX in a VirtualBox VM ( Ubuntu Desktop 16.04 LTS ).
The described issue persists. If I visit localhost:8008 the matrix web interface is shown ( URL changes to 'localhost:8008/_matrix/client/#/login' ) but if I visit 'matrix.whatever.private' ( as configured in the NGINX reverse proxy configuration ) only the welcome to NGINX page is shown.
If I type in manually 'matrix.whatever.private/_matrix' then the following message shows up:
'Method Not Allowed
Your browser approached me (at /_matrix) with the method "GET". I only allow the method HEAD here.'
Now I do not understand where the request using NGINX goes wrong.

Can you please explain how to configure the VirtualBox VM / remote-server correctly to get the reverse proxy working.
Please explain every step carefully because my experience with server administration is rather sparse.

### Steps to reproduce
I configured the VM as described below:

FYI: the VM has the hostname developer-VirtualBox which is not equal to the desired domain name matrix is accessed, on the remote server NGINX handles multiple domain-names ( at least two ).

//> Install matrix synapse ( as described at https://github.com/matrix-org/synapse )
sudo apt-get update
sudo apt-get upgrade
sudo apt-get install build-essential python2.7-dev libffi-dev python-pip python-setuptools sqlite3 libssl-dev python-virtualenv libjpeg-dev libxslt1-dev
virtualenv -p python2.7 ~/.synapse
source ~/.synapse/bin/activate
pip install --upgrade pip
pip install --upgrade setuptools
pip install https://github.com/matrix-org/synapse/tarball/master

//> Configure matrix synapse
cd ~/.synapse
python -m synapse.app.homeserver --server-name matrix.whatever.private --config-path homeserver.yaml --generate-config --report-stats=no

//> Register domain matrix.whatever.private in /etc/hosts
sudo nano /etc/hosts

//> Add the following lines to matrix.conf
127.0.1.1      matrix.whatever.private
// --

//> Reboot here... because /etc/hosts was changed
reboot

//> Start matrix synapse 
cd ~/.synapse
source ~/.synapse/bin/activate
bin/synctl start

//> Install reverse proxy
sudo apt-get install nginx

//> Configure reverse proxy
cd /etc/nginx/conf.d
sudo nano matrix.conf

//> Add the following lines to matrix.conf ( as described at https://github.com/matrix-org/synapse#using-a-reverse-proxy-with-synapse )
server {
    listen 80;
    server_name matrix.whatever.private;
    location /_matrix {
    proxy_pass http://localhost:8008
        proxy_set_header X-Forwarded-For $remote_addr;
    }
}
// --

//> Restart the reverse proxy
sudo systemctl restart nginx

### Version information

- **Homeserver**:  Yes
- **Version**:        Synapse/0.26.0
- **Install method**: git clone/pip      
- **Platform**:       Ubuntu Server 16.04 LTS ( 1und1.de Server ), Ubuntu Desktop 16.04 LTS ( VirtualBox VM )

**Comment by turt2live** [comment]
This is almost certainly a configuration problem within nginx. The people in [#matrix:matrix.org](https://matrix.to/#/#matrix:matrix.org) or [#synapse-community:matrix.org](https://matrix.to/#/#synapse-community:matrix.org) may be able to help out seeing as the issue tracker here is more for bug reports.

**Comment by neilisfragile** [comment]
+1 thanks @turt2live closing

@Dave-At-GitHub I couldn't see anything obvious on first glance, but the community rooms @turt2live references are great places to get help.

**Comment by Dave-At-GitHub** [comment]
You are right, it is not a bug so to speak, but I was referencing the documentation of this github project ( readme.rst ) where the example NGINX configuration is mentioned ( which does not seem to work ). Could you please update this, because it seems obvious to look at the project documentation when looking for information first. I'll ask for help in the community rooms and come back later ( when NGINX works as expected ) if that is ok. :)

**Comment by neilisfragile** [comment]
The example works as described, but in your case nginx is not picking up your config.

Since you are using Ubuntu you'll want to store the config in /etc/nginx/site-available and link through to it from /etc/nginx/site-enabled (conf.d would work for CentOS but to my knowledge not Ubuntu)

**Comment by Dave-At-GitHub** [comment]
I've updated the issue ( see section ###Edit### ).

**Comment by neilisfragile** [comment]
Thanks @Dave-At-GitHub you really shouldn't need to add additional rules to matrix.conf

I can't reproduce based on your description, though am assuming that 

//> Add the following lines to matrix.conf
127.0.1.1 matrix.whatever.private

is a typo in the description rather than a typo on the box?

**Comment by Dave-At-GitHub** [comment]
@neilisfragile: Yes it was a typo the line was meant to be added to the **/etc/hosts** file ( only required if you are using a domain name that is not publicly known ).
The original post is corrected now.
Sorry for the typo...

**Comment by Dave-At-GitHub** [comment]
Now everything should be up-to-date, the title is now a bit more descriptive and the complete nginx configuration is at the bottom of the original post. Can you update the section [Using a reverse proxy with Synapse](https://github.com/matrix-org/synapse#id38) and add the lines:

location = / { # Tells nginx to change the URL from https://your-domain-name.org to https://your-domain-name.org/_matrix/client/#/login and redirects correctly to the matrix-angular-sdk web-client
return 301 _matrix/client/#/login;
}

Or should I create a pull request for this?

**Comment by neilisfragile** [comment]
I can't reproduce your bug, it really shouldn't be necessary to include a 301. Other install guides are consistent with the matrix.org guide e..g https://www.digitalocean.com/community/tutorials/how-to-install-matrix-synapse-on-ubuntu-16-04

Could there be any other missing state or context on your nginx config?

**Comment by Dave-At-GitHub** [comment]
This issue regards to the web-interface [matrix-angular-sdk](https://github.com/matrix-org/matrix-angular-sdk) which seems to be installed only with the git clone/pip installation, the package-manager installation does not include the web-client at all. So the problem that nginx is not displaying the client in FireFox only can reproduced with the git clone/pip installation.

Regarding the 301 redirect: I do not know another easy method to tell the reverse proxy to extend the URL as required for the web-client ( otherwise you only will see the NGINX welcome page ).

**Comment by neilisfragile** [comment]
Okay, the initial report sounded like you were poking the api directly and were unable to reach synapse. It could be a problem with the matrix-angular-sdk but again this doesn't seem to be a general problem given other users of the client. 

To rule out the bundled client for sure, do you have the same problem when you point Riot at it?

**Comment by Dave-At-GitHub** [comment]
Yeah sorry for the confusion.
Riot works in all cases fine ( git clone / pip installation with or without additional nginx configuration, package-manager installation with or without additional nginx configuration ).

I mentioned the issue, because internally we are using matrix-synapse with Riot, but we also have external members which may use the ( pre-installed ) matrix-angular-sdk because the may do not install Riot on their devices.
First we installed matrix-synapse and used it without a reverse proxy, in that case the matrix-angular-sdk client worked as expected out-of-the-box ( the URL extended automatically and the client showed up ). But afterwards we moved the installation behind a reverse proxy ( to replace the extra port specification in the URL by a subdomain name ), after we did this the matrix-angular-sdk client did no longer show up...
That is the reason why I placed the issue here.
